riscv/decode_macros.h

```diff
@@ -321,3 +321,35 @@ inline long double to_f(float128_t f) { long double r; memcpy(&r, &f, sizeof(r))
   reg_t h##field = get_field(STATE.henvcfg->read(), HENVCFG_##field)
 
 #endif
+
+#define software_check(x, tval) (unlikely(!(x)) ? throw trap_software_check_fault(tval) : (void) 0)
+#define ZICFILP_READ_XLPE(v, prv) \
+  ({ \
+    reg_t csr_val, mask; \
+    if (p->extension_enabled('S')) { \
+      if (v) { \
+        switch (prv) { \
+        case PRV_U: csr_val = STATE.senvcfg->read(); mask = SENVCFG_LPE; break; \
+        case PRV_S: csr_val = STATE.henvcfg->read(); mask = HENVCFG_LPE; break; \
+        default: abort(); \
+        } \
+      } else { \
+        switch (prv) { \
+        case PRV_U: csr_val = STATE.senvcfg->read(); mask = SENVCFG_LPE; break; \
+        case PRV_S: csr_val = STATE.menvcfg->read(); mask = MENVCFG_LPE; break; \
+        case PRV_M: csr_val = STATE.mseccfg->read(); mask = MSECCFG_MLPE; break; \
+        default: abort(); \
+        } \
+      } \
+    } else { \
+      switch (prv) { \
+      case PRV_U: csr_val = STATE.menvcfg->read(); mask = MENVCFG_LPE; break; \
+      case PRV_M: csr_val = STATE.mseccfg->read(); mask = MSECCFG_MLPE; break; \
+      default: abort(); \
+      } \
+    } \
+    get_field(csr_val, mask); \
+  })
+#define ZICFILP_EXPECT_LP(reg_num) \
+  (((reg_num) != 1 && (reg_num) != 5 && (reg_num) != 7) ? \
```

Comment:
The magic numbers here need some explanation...I assume this is from the spec?

Reply:
could use the spec name for the macro - ZICFILP_IS_LP_EXPECTED. Could add the extension_enabled qualifier here.

Reply:
@scottj97 Yeah, the numbers come from the spec, which looks like this:
>   is_lp_expected = ( (JALR || C.JR || C.JALR) &&
>                     (rs1 != x1) && (rs1 != x5) && (rs1 != x7) ) ? 1 : 0;

Since the plain numbers just lay themselves out like that in the spec, I suppose for someone looking into this part of Zicfilp-specific code, it should not be hard to get the connection to spec. I can add a comment referencing the spec to clarify where the numbers came from, if this could address your concern?

-----

@ved-rivos I agree `ZICFILP_IS_LP_EXPECTED` is a naturally better name. But I think putting those test conditions you mentioned in this macro would lead to some drawbacks. Consider the macro containing the conditions:
```c++
#define ZICFILP_IS_LP_EXPECTED(reg_num) (has_zicfilp && xLPE && reg_num == 1/5/7 ? LP_EXPECTED : NO_LP_EXPECT)
```
This would cause the use site to become:
```c++
STATE.elp = ZICFILP_IS_LP_EXPECTED(rs1);
```
without any mention of testing whether Zicfilp is enabled, which
1. By my own standard, is a bit confusing to readers who would expect an `extension_enabled()` test to mark out optional behaviors of instructions, and
2. Forces `elp` update unnecessarily.

Due to these concerns, I think keeping the current content of `ZICFILP_EXPECT_LP` is better for readability and performance. What do you say?